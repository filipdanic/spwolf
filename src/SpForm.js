import React from 'react';
import PropTypes from 'prop-types';
import memoize from 'fast-memoize';
import {
  getHash,
  getDiff,
  getInitialState,
  flattenSections as flattenSections_,
  flattenValidationRules,
  formVisibilityDepMap,
  getValidationFeedback,
  formOnChangeDepList
} from './utils';

const flattenSections = memoize(flattenSections_);

export const FormHoC = ({ componentMap, wrappers }) => {
  class SpWolfForm extends React.Component {
    constructor(props) {
      super(props);
      this.state = getInitialState(props);
      this.handleChange_ = props.debugOnChange ? this.handleChangeWithDebug : this.handleChange;
    }

    componentDidMount() {
      const { specs } = this.props;
      const { defineStateGetter } = this.context;
      defineStateGetter && defineStateGetter(this.getFormState);
      const { conditionalFields = [] } = specs;
      this.conditionalFields = conditionalFields;
      const elements = flattenSections(specs.sections);
      this.validationFeedbackRules = flattenValidationRules(elements);
      this.visibilityMap = formVisibilityDepMap(elements);
      this.elementsWithOnChangeReset = formOnChangeDepList(elements);
      this.calculateAllConditionalFields();
    }

    getFormState = () => ({
      initialState: this.state.initialState,
      state: this.state.entityState,
      diff: getDiff(this.state.initialState, this.state.entityState),
    });

    updateCanSubmitForm = (canSubmitForm) => {
      this.setState({ canSubmitForm });
      this.props.onCanSubmitFormChange && this.props.onCanSubmitFormChange(canSubmitForm);
    };

    validateForm = () => {
      const { entityState } = this.state;
      const { specs } = this.props;
      const elements = flattenSections(specs.sections);

      for (let i = 0; i < elements.length; i += 1) {
        const element = elements[i];
        const value = entityState[element.name];
        let existsInFormContext = true;
        if (element.existsIf) {
          existsInFormContext = element.existsIf
            .reduce((acc, cond) => acc && entityState[cond], true);
        }

        if (existsInFormContext) {
          if (element.validationFeedbackRules) {
            const feedback = this.validationFeedbackRules[element.name]
              .find(rule =>
                getValidationFeedback(
                  element.name,
                  entityState,
                  rule.condition,
                  rule.validateWith
                ));
            if (feedback && feedback.type === 'error') {
              this.updateCanSubmitForm(false);
              break;
            }
          } else if (element.required) {
            if ((value === undefined || value === '')) {
              this.updateCanSubmitForm(false);
              break;
            }
          }
        }
        if (i === elements.length - 1) {
          const { diff } = this.getFormState();
          if (Object.keys(diff).length > 0) {
            this.updateCanSubmitForm(true);
          } else {
            this.updateCanSubmitForm(false);
          }
        }
      }
    };

    validateField = (field, checkOnChangeOnly) => {
      if (this.validationFeedbackRules[field]) {
        const { validationFeedback } = this.state;
        const feedback = this.validationFeedbackRules[field]
          .find(rule =>
            (checkOnChangeOnly ? rule.checkOnChange : true) &&
            getValidationFeedback(
              field,
              this.state.entityState,
              rule.condition,
              rule.validateWith
            ));

        validationFeedback[field] = {
          hasFeedback: feedback !== undefined,
          type: (feedback || {}).type,
          label: (feedback || {}).label
        };
        this.setState({ validationFeedback });
      }
    };

    handleFieldBlur = (field) => {
      this.validateField(field);
    };

    evaluateAsyncDep = (fn, dependants) => {
      const state = dependants.reduce((acc, _) =>
        Object.assign({}, acc, { [_]: this.state.entityState[_] }), {});
      const hash = getHash(state);
      if (this.state.cachedAsyncFields[hash]) {
        return this.state.cachedAsyncFields[hash];
      }
      try {
        fn(state).then((result) => {
          const cache = this.state.cachedAsyncFields;
          cache[hash] = result;
          this.setState({ cachedAsyncFields: cache });
        });
      } catch (err) {
        // TODO
      }
      return { asyncPending: true };
    };

    calculateAllConditionalFields = () => {
      const prev = Object.assign({}, this.state.entityState);
      const next = this.conditionalFields.reduce((state, field) => {
        return Object.assign({}, state, { [field.name]: field.fn(this.state.entityState) });
      }, prev);
      this.setState({ entityState: next }, () => {
        this.validateForm();
      });
    };

    handleChangeWithDebug = ({ key, value }) => {
      const withReset = this.elementsWithOnChangeReset[key] ? {
        [this.elementsWithOnChangeReset[key]]: undefined,
      } : {};
      console.group('spForm: handleChange');
      console.log(`handleChange called for "${key}" with value: `, value);
      const entityState = Object.assign(
        {},
        this.state.entityState,
        withReset,
        { [key]: value }
      );
      if (this.elementsWithOnChangeReset[key]) {
        console.log(`%cThe key ${this.elementsWithOnChangeReset[key]} will be reset.`, 'color: red;');
      }
      console.log('Diff: ', getDiff(this.state.entityState, entityState));
      console.groupEnd();
      this.setState({ entityState }, () => {
        this.dispatchChangedKey(key);
        this.validateField(key, true);
      });
    };

    handleChange = ({ key, value }) => {
      const withReset = this.elementsWithOnChangeReset[key] ? {
        [this.elementsWithOnChangeReset[key]]: undefined,
      } : {};
      const entityState = Object.assign(
        {},
        this.state.entityState,
        withReset,
        { [key]: value }
      );
      this.setState({ entityState }, () => {
        this.dispatchChangedKey(key);
        this.validateField(key, true);
      });
    };

    dispatchChangedKey = (key) => {
      clearTimeout(this.timer_);
      const prev = Object.assign({}, this.state.entityState);
      const next = this.conditionalFields.reduce((state, field) => {
        if (field.dependsOn.some(dependant => dependant === key)) {
          const value = field.fn(this.state.entityState);
          let inferredUpdates = {};
          if (value === false) {
            inferredUpdates = (this.visibilityMap[field.name] || [])
              .reduce((acc, _) =>
                Object.assign({}, acc, { [_]: undefined }), {});
          }
          return Object.assign({}, state, inferredUpdates, { [field.name]: value });
        }
        return state;
      }, prev);
      this.setState({ entityState: next }, () => {
        this.timer_ = setTimeout(() => this.validateForm(), 250);
      });
    };

    render() {
      const { specs, sectionProps } = this.props;
      const { entityState, validationFeedback } = this.state;
      const Form = wrappers.form;
      const Section = wrappers.section;
      return (
        <Form>
          {specs.sections.map((section, i) =>
            <Section
              key={`section-${i}`}
              index={i}
              meta={section.meta || {}}
              {...sectionProps}
            >
              {section.elements.map(({
                 fieldType,
                 existsIf = [],
                 disabledIf,
                 asyncEval,
                 isPresentationalElement,
                 ...otherProps
               }, j) => {
                const FormFieldComponent = componentMap[fieldType];
                const isVisible = existsIf.reduce((acc, cond) => acc && entityState[cond], true);
                const isDisabled = entityState[disabledIf] || false;

                if (!isVisible) {
                  return null;
                }

                if (isPresentationalElement) {
                  const presentationalData = (otherProps.dependsOn || [])
                    .reduce((acc, field) =>
                      Object.assign({}, acc, { [field]: entityState[field] }), {});
                  return <FormFieldComponent key={`field-${j}`} data={presentationalData} />;
                }

                const inferredProps = {};
                if (asyncEval && Array.isArray(asyncEval) && asyncEval.length > 0) {
                  asyncEval.forEach((asyncSpec) => {
                    inferredProps[asyncSpec.key] = this.evaluateAsyncDep(...asyncSpec.config);
                  });
                }
                return (
                  <FormFieldComponent
                    className="margin-bottom-12"
                    key={`field-${j}`}
                    onChange={this.handleChange_}
                    value={entityState[otherProps.name]}
                    onFieldBlur={this.handleFieldBlur}
                    validationFeedback={validationFeedback[otherProps.name]}
                    disabled={isDisabled}
                    {...otherProps}
                    {...inferredProps}
                  />
                );
                })}
            </Section>)}
        </Form>
      );
    }
  }

  SpWolfForm.propTypes = {
    onCanSubmitFormChange: PropTypes.func,
    debugOnChange: PropTypes.bool,
    specs: PropTypes.shape({
      sections: PropTypes.arrayOf(PropTypes.shape({
        meta: PropTypes.object,
        elements: PropTypes.arrayOf(PropTypes.shape({
          name: PropTypes.string.isRequired,
          fieldType: PropTypes.string.isRequired,
          required: PropTypes.bool,
          isPresentationalElement: PropTypes.bool,
          dependsOn: PropTypes.arrayOf(PropTypes.string),
          existsIf: PropTypes.arrayOf(PropTypes.string),
          disabledIf: PropTypes.string,
          onChangeReset: PropTypes.string,
          validationFeedbackRules: PropTypes.arrayOf(PropTypes.shape({
            type: PropTypes.string.isRequired,
            condition: PropTypes.func.isRequired,
            label: PropTypes.string,
            checkOnChange: PropTypes.bool
          }))
        })).isRequired
      })).isRequired,
      conditionalFields: PropTypes.arrayOf(PropTypes.shape({
        name: PropTypes.string.isRequired,
        fn: PropTypes.func.isRequired,
        dependsOn: PropTypes.arrayOf(PropTypes.string)
      }))
    }).isRequired
  };

  SpWolfForm.defaultProps = {
    onCanSubmitFormChange: undefined,
    debugOnChange: false
  };

  SpWolfForm.contextTypes = {
    defineStateGetter: PropTypes.func
  };

  return SpWolfForm;
};

export const connect = (Component) => {
  class Connected extends React.Component {
    state = { getFormState: undefined };

    defineStateGetter = (cb) => {
      this.setState({ getFormState: cb });
    };

    getChildContext() {
      return {
        defineStateGetter: this.defineStateGetter
      };
    }

    render() {
      return (
        <Component getFormState={this.state.getFormState} {...(this.props || {})} />
      );
    }
  }

  Connected.childContextTypes = {
    defineStateGetter: PropTypes.func
  };
  return Connected;
};
