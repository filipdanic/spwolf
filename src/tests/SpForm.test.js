import React from 'react'
import renderer from 'react-test-renderer'
import { FormHOC } from '../SpForm'

const TextField = ({ label, value, onChange, required, name }) => (
  <label htmlFor={name}>
    {label}
    <input
      name={name}
      id={name}
      required={required}
      type='text'
      value={value}
      onChange={onChange}
    />
  </label>
)

const SectionWrap = ({ children, meta }) => (
  <div>
    <h2 style={meta.bold ? { fontWeight: 700 } : {}}>
      {meta.title}
    </h2>
    <div>
      {children}
    </div>
  </div>
)
const FormWrap = ({ children }) => <div>{children}</div>
const specs = {
  sections: [{
    elements: [{
      name: 'foo',
      fieldType: 'text',
      label: 'Foo',
      required: true
    }],
    meta: { title: 'Foo', bold: true }
  }, {
    elements: [{
      name: 'bar',
      fieldType: 'text',
      label: 'Bar'
    }, {
      name: 'barConditional',
      fieldType: 'text',
      label: 'Bar Conditional',
      disabledIf: 'isBarInvalid'
    }],
    meta: { title: 'Bar' }
  }],
  conditionalFields: [{
    name: 'isBarInvalid',
    fn: state => !(state.bar) || state.bar.length > 20,
    dependsOn: ['bar']
  }]
}

describe('SpWolf', () => {
  it('renders SpWolf component correctly after initializing via FormHOC', () => {
    const SpWolf = FormHOC({
      componentMap: {
        text: TextField
      },
      wrappers: {
        form: FormWrap,
        section: SectionWrap
      }
    })
    const tree = renderer.create(<SpWolf
      initialState={{ foo: 'foofoo' }}
      specs={specs}
      onCanSubmitFormChange={val => console.log(val)}
    />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
