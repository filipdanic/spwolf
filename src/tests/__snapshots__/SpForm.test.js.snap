// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpWolf renders SpWolf component correctly after initializing via FormHOC 1`] = `
<div>
  <div>
    <h2
      style={
        Object {
          "fontWeight": 700,
        }
      }
    >
      Foo
    </h2>
    <div>
      <label
        htmlFor="foo"
      >
        Foo
        <input
          id="foo"
          name="foo"
          onChange={[Function]}
          required={true}
          type="text"
          value="foofoo"
        />
      </label>
    </div>
  </div>
  <div>
    <h2
      style={Object {}}
    >
      Bar
    </h2>
    <div>
      <label
        htmlFor="bar"
      >
        Bar
        <input
          id="bar"
          name="bar"
          onChange={[Function]}
          required={undefined}
          type="text"
          value={undefined}
        />
      </label>
      <label
        htmlFor="barConditional"
      >
        Bar Conditional
        <input
          id="barConditional"
          name="barConditional"
          onChange={[Function]}
          required={undefined}
          type="text"
          value={undefined}
        />
      </label>
    </div>
  </div>
</div>
`;
